src/keyring/ed25519/signing_key.rs

```diff
@@ -6,70 +7,120 @@ use signature::Signer;
 type SigningKeyBytes = [u8; SigningKey::BYTE_SIZE];
 
 /// Ed25519 signing key.
-#[derive(Clone, Debug)]
-pub struct SigningKey(ed25519_consensus::SigningKey);
+pub enum SigningKey {
+    /// Ed25519 signing key.
+    Ed25519(ed25519_consensus::SigningKey),
+    /// Ed25519 expanded signing key.
+    Ed25519Expanded(ExpandedSecretKey),
+}
 
 impl SigningKey {
     /// Size of an encoded Ed25519 signing key in bytes.
     pub const BYTE_SIZE: usize = 32;
 
+    /// Size of an Ed25519 expanded signing key in bytes.
+    pub const EXPANDED_BYTE_SIZE: usize = 64;
+
+    /// Size of an Ed25519 public key in bytes.
+    pub const PUBLIC_KEY_BYTE_SIZE: usize = 32;
+
     /// Borrow the serialized signing key as bytes.
     pub fn as_bytes(&self) -> &SigningKeyBytes {
-        self.0.as_bytes()
+        match &self {
+            SigningKey::Ed25519(signing_key) => signing_key.as_bytes(),
+            SigningKey::Ed25519Expanded(_) => panic!("unexpected expanded signing key"),
+        }
     }
 
     /// Get the verifying key for this signing key.
     pub fn verifying_key(&self) -> VerifyingKey {
-        VerifyingKey(self.0.verification_key())
+        match &self {
+            SigningKey::Ed25519(signing_key) => VerifyingKey(signing_key.verification_key()),
+            SigningKey::Ed25519Expanded(signing_key) => Into::<ed25519_dalek::VerifyingKey>::into(signing_key).as_bytes().as_slice().try_into().unwrap(),
```

Comment:
This is a trick so I don't have to reimplement the verifying key for expanded secret keys separately:
1. Convert the ExpandedSecretKey into ed25519-dalek's VerifyingKey (trait implemented in ed25519-dalek)
2. Get the bytes from the VerifyingKey, turn them into a slice
3. Use the trait defined in ed25519-consensus to turn a slice of bytes into an ed25519-consensus VerifyingKey.

The verifying key implementation already falls back to this function so this way I could implement both types of verifying keys in one go.